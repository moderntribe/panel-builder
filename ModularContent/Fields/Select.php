<?php


namespace ModularContent\Fields;

use ModularContent\Panel;


/**
 * Class Select
 *
 * @package ModularContent\Fields
 *
 * A select box.
 */
class Select extends Field {

	protected $options                = [];
	protected $options_cache          = NULL;
	protected $layout                 = 'compact';
	protected $global_options         = false;
	protected $enable_fonts_injection = false;

	/**
	 * @param array $args
	 */
	public function __construct( $args = [] ) {
		$this->check_layout( $args );

		$this->defaults['options']                = $this->options;
		$this->defaults['layout']                 = $this->layout;
		$this->defaults['global_options']         = $this->global_options;
		$this->defaults['enable_fonts_injection'] = $this->enable_fonts_injection;
		parent::__construct( $args );
	}

	protected function check_layout( $args ) {
		if ( isset( $args['layout'] ) && $args['layout'] !== 'compact' && $args['layout'] !== 'full' && $args['layout'] !== 'inline' && $args['layout'] !== 'horizontal' ) {
			throw new \LogicException( 'Layout argument can only be "compact" or "full", you passed "' . $args['layout'] . '".' );
		}
	}

	protected function get_options() {
		if ( isset( $this->options_cache ) ) {
			return $this->options_cache;
		}
		if ( empty( $this->options ) ) {
			return array();
		}
		if ( is_callable( $this->options ) ) {
			$this->options_cache = call_user_func( $this->options );
		} else {
			$this->options_cache = $this->options;
		}
		return $this->options_cache;
	}

	public function get_blueprint() {
		$blueprint            = parent::get_blueprint();
		$options              = $this->get_options();
		$blueprint['options'] = [];
		foreach ( $options as $key => $label ) {
			$blueprint['options'][] = [
				'label' => $label,
				'value' => (string) $key, // cast to string so react-select has consistent types for comparison
			];
		}
		$blueprint['layout']                 = $this->layout;
		$blueprint['global_options']         = $this->global_options;
		$blueprint['enable_fonts_injection'] = $this->enable_fonts_injection;
		return $blueprint;
	}

	/**
	 * Massage submitted data before it's saved.
	 *
	 * @param mixed $data
	 *
	 * @return string
	 */
	public function prepare_data_for_save( $data ) {
		$data = (string) $data;
		if ( strlen( $data ) === 0 && (string) $this->default ) {
			$data = (string) $this->default;
		}
		return $data;
	}
} 