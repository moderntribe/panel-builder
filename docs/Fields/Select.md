## Select

A select box.

### Parameters

||||
|---|---|---|
| `label`       | `string` | The __() translated label for this field       |
| `name`        | `string` | The unique name for this field                 |
| `description` | `string` | The __() translated description for this field |
| `options` | `array` | An array of key => value pairs, where the `key` is the option name and the `value` is the option label |
| `global_options` | `string` | Used _only_ if you need to load options based on a separate, localized JSON object. (Usually not needed) |
| `enable_fonts_injection` | `bool` | If `true`, allows this select to have fonts injected into it to render font previews (Must be handled in the JS) |

### Example

```php
$field = new Select( [
    'label' => __('Pick One'),
    'name' => 'my-field',
    'description' => __( 'Pick the thing that you pick', 'tribe' )
    'options' => [
        'first' => __( 'The First Option', 'tribe' ),
        'second' => __( 'The Second Option', 'tribe' ),
    ]
] );
```