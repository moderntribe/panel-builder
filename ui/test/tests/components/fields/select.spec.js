/* global expect b:true */

import React from 'react';
import { shallow, mount } from 'enzyme';
import Select from 'components/fields/select';
import renderer from 'react-test-renderer';

// test data
const OPTIONS1 = [
	{
		label: 'Standard',
		value: 'standard',
	},
	{
		label: 'Cards',
		value: 'cards',
	},
	{
		label: 'Full',
		value: 'full',
	},
];
const OPTIONS2 = [
	{
		label: 'Orange',
		value: '1',
	},
	{
		label: 'Raspberry',
		value: '2',
	},
];

describe('Select field', () => {
	it('renders correctly', () => {
		const tree = renderer
			.create(<Select description="This is a test description" label="Some test label" options={OPTIONS1} />)
			.toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('renders as a <div>', () => {
		const wrapper = shallow(<Select />);
		expect(wrapper.type()).toEqual('div');
	});

	it('has no radio buttons without properties.', () => {
		const wrapper = shallow(<Select />);
		expect(wrapper.find('input').length).toEqual(0);
	});

	it('has a label', () => {
		const wrapper = shallow(<Select />);
		expect(wrapper.find('label').length).toEqual(1);
	});

	it('allows us to set and edit props', () => {
		const wrapper = mount(<Select description="This is a test description" label="Some test label" options={OPTIONS1} name="layout" default="cards" />);
		expect(wrapper.props().description).toEqual('This is a test description');
		wrapper.setProps({ description: 'Change description for test' });
		expect(wrapper.props().description).toEqual('Change description for test');

		expect(wrapper.props().label).toEqual('Some test label');
		wrapper.setProps({ label: 'Change label for test' });
		expect(wrapper.props().label).toEqual('Change label for test');

		expect(wrapper.props().name).toEqual('layout');
		wrapper.setProps({ name: 'fruit' });
		expect(wrapper.props().name).toEqual('fruit');

		expect(wrapper.props().default).toEqual('cards');
		wrapper.setProps({ default: '1' });
		expect(wrapper.props().default).toEqual('1');

		expect(wrapper.props().options).toEqual(OPTIONS1);
		wrapper.setProps({ options: OPTIONS2 });
		expect(wrapper.props().options).toEqual(OPTIONS2);
	});
});
