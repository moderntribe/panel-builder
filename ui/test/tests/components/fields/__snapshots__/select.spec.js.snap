exports[`Select field renders correctly 1`] = `
<div
  className="field compact panel-field panel-conditional-field">
  <label
    className="label panel-field-label">
    Some test label
    <span
      className="icon"
      onBlur={undefined}
      onClick={undefined}
      onContextMenu={undefined}
      onFocus={undefined}
      onMouseDown={undefined}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onTouchStart={undefined} />
  </label>
  <div
    className="Select Select--single is-clearable is-searchable"
    style={undefined}>
    <div
      className="Select-control"
      onKeyDown={[Function]}
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      style={undefined}>
      <span
        className="Select-multi-value-wrapper"
        id="react-select-2--value">
        <div
          className="Select-placeholder">
          Select...
        </div>
        <div
          className="Select-input"
          style={
            Object {
              "display": "inline-block",
            }
          }>
          <input
            aria-activedescendant="react-select-2--value"
            aria-describedby={undefined}
            aria-expanded="false"
            aria-haspopup="false"
            aria-label={undefined}
            aria-labelledby={undefined}
            aria-owns=""
            className={undefined}
            id={undefined}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            role="combobox"
            style={
              Object {
                "boxSizing": "content-box",
                "width": "5px",
              }
            }
            tabIndex={undefined}
            value="" />
          <div
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "scroll",
                "position": "absolute",
                "top": 0,
                "visibility": "hidden",
                "whiteSpace": "pre",
              }
            }>
            
          </div>
        </div>
      </span>
      <span
        className="Select-arrow-zone"
        onMouseDown={[Function]}>
        <span
          className="Select-arrow"
          onMouseDown={[Function]} />
      </span>
    </div>
  </div>
</div>
`;
